### PR TITLE
Prevent flickering in nushell due to FTCS marks

### LINKED ISSUE #13710: [1.15+] Flushing FTCS sequences causes flickering in nushell 0.66+
### Windows Terminal version

1.15.2003.0

### Windows build number

Microsoft Windows [Version 10.0.22000.795]

### Other Software

Nushell v0.66+

### Steps to reproduce

Hi, I work on a cross-platform shell called [Nushell](https://github.com/nushell/nushell) and [we've noticed a rendering regression in Windows Terminal Preview v1.15](https://github.com/nushell/nushell/issues/6214).

Nushell's prompt now flickers on every keystroke. The following recording illustrates the problem (but it's a little worse in real life, the recording didn't capture every flicker):

https://user-images.githubusercontent.com/26268125/183734387-10477dd6-220c-4f7d-8b21-f7737736d0ee.mp4

To reproduce this, run any version of Nushell from v0.66 or later and start typing. Nushell can be downloaded from [our GitHub releases](https://github.com/nushell/nushell/releases/tag/0.66.2) or installed with Scoop (`scoop install nushell`) or winget (`winget install nushell`). `nu.exe` is the executable for Nushell.

## Details / Investigation

The flickering does _not_ occur in Windows Terminal v1.14.1963.0.

We bisected Nushell's code and found that the flickering did not occur prior to [this PR](https://github.com/nushell/nushell/pull/6087) which added some scroll mark ANSI escape codes for VS Code. WT v1.15 Preview [added experimental support for scroll marks](https://devblogs.microsoft.com/commandline/windows-terminal-preview-1-15-release/#scroll-marks-experimental); maybe this is a bug in the experimental scroll mark code?

The flickering occurs even with the scroll mark settings (`experimental.autoMarkPrompts` and `experimental.showMarksOnScrollbar`) set to false.

If there's anything else we can do to help with investigation or reproduction, just let me know!

### PR DESCRIPTION
**Pull request by zadjii-msft**
Tl;dr: Conpty would flush a frame whenever it encountered a FTCS mark. Combine that with the whole-line redrawing that nushell does, and the Terminal would get the prompt in two frames instead of	one, causing a slight flickering. This fixes that by rendering the frame, but not flushing to the pipe when we encounter one of these sequences. 

Closes #13710 

A complication here: there are some sequences that we passthrough _immediately_ when we encounter them. For example, `\x1b[ 2q`. we need to also not flush when we encounter one of these sequences. nushell emits one of these as a part of the prompt, and that would force the buffered frame to get written _anyways_, before writing that to the pipe.

### CODE REVIEW

File: src/renderer/vt/XtermEngine.cpp [hunk 1]
```diff
@@ -534,10 +534,16 @@ CATCH_RETURN();
     // decided to pass through would have gotten buffered here until the next
     // actual frame is triggered.
     //
-    // To fix this, flush here, so this string is sent to the connected terminal
-    // application.
+    // Originally, we would _Flush()  here, so this string (and anything buffered) is sent to the connected
+    // terminal application. As of GH#13710, we won't. We'll just buffer the string.
```
Comment (DHowett) [inline src/renderer/vt/XtermEngine.cpp:538]: nit: same comment as on another PR - signposts

Comment (DHowett) [reply src/renderer/vt/XtermEngine.cpp:538]: I guess what I mean is, "make the comment explain why the code is doing what it's doing, rather than explaining why we added and removed the code ourselves"

**Review by DHowett — CHANGES_REQUESTED** [review]
I might need you to walk me through how the "flush to terminal" callback becoming "paint and then also flush" _actually prevents excess flushes_ :smile:

File: src/renderer/vt/paint.cpp [hunk 2]
```diff
@@ -538,7 +553,8 @@ using namespace Microsoft::Console::Types;
     // Write the actual text string. If we're using a soft font, the character
     // set should have already been selected, so we just need to map our internal
     // representation back to ASCII (handled by the _WriteTerminalDrcs method).
-    if (_usingSoftFont) [[unlikely]]
+    if (_usingSoftFont)
+        [[unlikely]]
```
Comment (j4james) [inline src/renderer/vt/paint.cpp:557]: Not that it matters that much, but why was this changed? All previous uses of `[[unlikely]]` have been on the same line as the `if`, so it would be nice if we could remain consistent about this.

File: src/terminal/parser/OutputStateMachineEngine.cpp [hunk 3]
```diff
@@ -168,6 +168,7 @@ bool OutputStateMachineEngine::ActionPassThroughString(const std::wstring_view s
     if (_pTtyConnection != nullptr)
     {
         const auto hr = _pTtyConnection->WriteTerminalW(string);
+
         LOG_IF_FAILED(hr);
```
Comment (j4james) [inline src/terminal/parser/OutputStateMachineEngine.cpp:172]: Again not important, but I'm curious whether there was a reason for this, or just the cat getting onto the keyboard?

File: src/renderer/vt/paint.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp:557]: I think this is my version of `clang-format` being overly aggressive 😓

File: src/terminal/parser/OutputStateMachineEngine.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/terminal/parser/OutputStateMachineEngine.cpp:172]: Honestly, couldn't tell you :p

**Comment by zadjii-msft** [comment]
@DHowett Okay I'm so sorry my handwriting is atrocious, but this is faster than trying to mermaid it

image

(imagine an "after" beneath the dotted line)

**Comment by zadjii-msft** [comment]
For real though:

### Before

```mermaid
sequenceDiagram
    participant API Thread
    participant Render Thread
    participant Terminal
    Note left of API Thread: {the start of the prompt}
    Note left of API Thread: ^[[2q
    activate API Thread
    opt Parser
    Note right of API Thread: StateMachine::ProcessString
    Note right of API Thread: AdaptDispatch -> failed
    Note right of API Thread: StateMachine::FlushToTerminal
    end
    opt XtermEngine
        Note right of API Thread: InvalidateFlush
        Note right of API Thread: [paint the frame]
        Note right of API Thread: WriteTerminalW
        Note right of API Thread: write "^[[2q"to _buffer
        Note right of API Thread: _Flush
        API Thread ->> Terminal: _buffer goes out the pipe
    end
    Note left of API Thread: {the rest of the prompt}
    Note left of API Thread: Unlock console
    API Thread --> -Render Thread: 
    activate Render Thread
        Note right of Render Thread: PaintFrame
        opt XtermEngine
            Note right of Render Thread: [paint the frame]
            Note right of Render Thread: _Flush
        end
        Render Thread ->> Terminal: _buffer goes out the pipe
    deactivate Render Thread
```

### After this PR
```mermaid
sequenceDiagram
    participant API Thread
    participant Render Thread
    participant Terminal
    Note left of API Thread: {the start of the prompt}
    Note left of API Thread: ^[[2q
    activate API Thread
        opt Parser
            Note right of API Thread: StateMachine::ProcessString
            Note right of API Thread: AdaptDispatch -> failed
            API Thread -->> Render Thread: NotifyPaintFrame
            Note right of API Thread: StateMachine::FlushToTerminal
        end
        opt XtermEngine
            Note right of API Thread: InvalidateFlush
            Note right of API Thread: [paint the frame]
            Note right of API Thread: WriteTerminalW
            Note right of API Thread: write "^[[2q"to _buffer
        end
    Note left of API Thread: {the rest of the prompt}
    Note left of API Thread: Unlock console
    API Thread --> -Render Thread: 
    activate Render Thread
        Note right of Render Thread: PaintFrame
        opt XtermEngine
            Note right of Render Thread: [paint the frame]
            Note right of Render Thread: _Flush
        end
        Render Thread ->> Terminal: _buffer goes out the pipe
    deactivate Render Thread
```

File: src/renderer/vt/XtermEngine.cpp [hunk 4]
```diff
@@ -534,10 +534,16 @@ CATCH_RETURN();
...
-    // To fix this, flush here, so this string is sent to the connected terminal
-    // application.
+    // Originally, we would _Flush()  here, so this string (and anything buffered) is sent to the connected
+    // terminal application. As of GH#13710, we won't. We'll just buffer the string.
+    //
+    // Now, we've added a NotifyPaintFrame to the state machine's
+    // FlushToTerminal callback. That callback will allow us to add this wstr to
+    // the buffer now, without pushing the entire buffer out to the pipe right
+    // now. The NotifyPaintFrame in that callback will ensure that a Flush
+    // definitely _does_ happen (after this whole string is processed).
```
Comment (lhecker) [inline src/renderer/vt/XtermEngine.cpp:544]: This last sentence is a little confusing.

File: src/renderer/vt/invalidate.cpp [hunk 5]
```diff
@@ -120,6 +120,12 @@ CATCH_RETURN();
         _circled = circled;
     }
 
+    // If we flushed for any reason other than circling, we don't need to push
+    // the buffer out on EndPaint. We're doing this because we encountered some
+    // VT sequence that required we sync the state of the buffers before passing
+    // it through.
```
Comment (lhecker) [inline src/renderer/vt/invalidate.cpp:126]: I also don't quite understand this sentence. The code does make sense to me: "don't flush on end when it's not circling, aka shuffling rows outside of the viewport", but I don't quite get how the comment fits with that... 🤔

**Review by lhecker — COMMENTED** [review]
The code does look good to me and I'd approve it instantly, but I'm not sure whether this is the best approach. From what I can tell, flushing during circling isn't necessarily at all and I believe it'd be better in general to instead only flush when the console API call finishes (i.e. when the entire user string has been processed), because this will avoid calling `WriteFile` for every line that's written. It should also solve the issue, while also improving ConPTY performance drastically.

**Review by lhecker — APPROVED** [review]

File: src/renderer/vt/paint.cpp [hunk 6]
```diff
@@ -94,7 +94,21 @@ using namespace Microsoft::Console::Types;
...
 
-    RETURN_IF_FAILED(_Flush());
+    // If this frame was triggered because we encountered a VT sequence which
+    // required the buffered state to get printed, we don't want to flush this
+    // frame to the pipe. That might result in us rendering half the output of a
+    // particular frame (as emitted by the client).
+    //
+    // Instead, we'll leave this frame in _buffer, and just keep appending to
+    // it as needed.
+    if (_noFlushOnEnd) [[unlikely]]
```
Comment (lhecker) [inline src/renderer/vt/paint.cpp:104]: FYI I don't think the `[[unlikely]]` is worth it here. Its implementation is super dumb in MSVC and causes any and all affected unlikely code to be moved to the end of the function, even if it's literally just a single `mov [rcx+1234], 0` instruction like here. I think it's still useful, but it seems like its usage is mostly restricted to a rather "nuanced" application, especially after observing the disassembly output.

**Review by DHowett — APPROVED** [review]
